Add validation for SQL queries in DataFrame.readSqlQuery

zaleslaw opened the pull request:
Fixes #498 

A validation function has been added to `DataFrame.readSqlQuery` to handle cases where an inappropriate SQL query is passed. 

The function verifies that the SQL query starts with "SELECT" and doesn't contain any other DDL, DML, or DCL operators to prevent data manipulation. 

Corresponding test cases were also added in h2Test to ensure that an `IllegalArgumentException` is thrown when the SQL query is incorrect.

Jolanrensen commented on dataframe-jdbc/src/main/kotlin/org/jetbrains/kotlinx/dataframe/io/readJdbc.kt:
```diff
@@ -199,6 +201,19 @@ public fun DataFrame.Companion.readSqlQuery(connection: Connection, sqlQuery: St
+        !forbiddenKeywords.any { normalizedSqlQuery.contains(it) }
+}
+
+private val forbiddenKeywords = setOf("INSERT", "UPDATE", "DELETE", "LOCK",
```
sorry, not so forbidden:
```sql
CREATE TABLE "ALTER"
(
    id SERIAL PRIMARY KEY,
    description TEXT
);
```
is valid

Jolanrensen replied:
so
```sql
SELECT * from "ALTER";
```
is too

zaleslaw replied:
Could you explain your comment? Do you mean that given example are passed the condition? Or should be accepted as an SQL queries?

Jolanrensen replied:
`SELECT * from "ALTER"` is a valid SQL query and should be accepted, however, since you check whether the query doesn't contain any of these keywords, it won't be accepted atm.

zaleslaw commented on dataframe-jdbc/src/main/kotlin/org/jetbrains/kotlinx/dataframe/io/readJdbc.kt:
```diff
@@ -199,6 +201,19 @@ public fun DataFrame.Companion.readSqlQuery(connection: Connection, sqlQuery: St
+    val normalizedSqlQuery = sqlQuery.trim().uppercase()
+
+    return normalizedSqlQuery.startsWith("SELECT") &&
+        !normalizedSqlQuery.contains(";") &&
```
Probably the idea is the opposite - I suggest accepting only without any ; - if it contains, it should be removed

Jolanrensen replied:
oh you're right, I got confused with the inversion.
Uhm, well in that case. Is a query necessarily invalid if it contains a `;` at the end?

zaleslaw commented on dataframe-jdbc/src/main/kotlin/org/jetbrains/kotlinx/dataframe/io/readJdbc.kt:
```diff
@@ -199,6 +201,19 @@ public fun DataFrame.Companion.readSqlQuery(connection: Connection, sqlQuery: St
+private fun isValid(sqlQuery: String): Boolean {
+    val normalizedSqlQuery = sqlQuery.trim().uppercase()
+
+    return normalizedSqlQuery.startsWith("SELECT") &&
```
My bad! Of course, you are right!

Jolanrensen replied:
haha still no constant?

zaleslaw commented:
@Jolanrensen I simplified a little bit of code and added some tests

> Tests were added to the h2Test to cover cases reported in issues https://github.com/Kotlin/dataframe/issues/494 and https://github.com/Kotlin/dataframe/issues/498 in the DataFrame. Furthermore, SQL query validation in DataFrame.readSqlQuery was softened to accept SELECT statements only, removing other restrictions that checked against DDL or DML or DCL operators. New test cases were introduced to ensure that unexpected SQL statements result in thrown IllegalArgumentExceptions, improving overall robustness.
